ambari-metrics/ambari-metrics-timelineservice/src/main/java/org/apache/ambari/metrics/core/timeline/upgrade/AbstractPhoenixMetricsCopier.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ambari.metrics.core.timeline.upgrade;
+
+import org.apache.ambari.metrics.core.timeline.PhoenixHBaseAccessor;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Set;
+
+public abstract class AbstractPhoenixMetricsCopier implements Runnable {
+  private static final Log LOG = LogFactory.getLog(AbstractPhoenixMetricsCopier.class);
+  private final Long startTime;
+  protected final FileWriter processedMetricsFile;
+  protected String inputTable;
+  protected String outputTable;
+  protected Set<String> metricNames;
+  protected PhoenixHBaseAccessor hBaseAccessor;
+
+  public AbstractPhoenixMetricsCopier(String inputTableName, String outputTableName, PhoenixHBaseAccessor hBaseAccessor,
+                                      Set<String> metricNames, Long startTime, FileWriter outputStream) {
+    this.inputTable = inputTableName;
+    this.outputTable = outputTableName;
+    this.hBaseAccessor = hBaseAccessor;
+    this.metricNames = metricNames;
+    this.startTime = startTime;
+    this.processedMetricsFile = outputStream;
+  }
+
+  public void run(){
+    long startTimer = System.currentTimeMillis();
+    if (metricNames == null || metricNames.isEmpty()) {
+      String query = String.format("SELECT * FROM %s WHERE SERVER_TIME > %s ORDER BY METRIC_NAME, SERVER_TIME", inputTable, startTime);
+      runPhoenixQueryAndAddToResults(query);
+    } else {
+      for (String metricName : metricNames) {
+//      TODO do batch selects
+        String query = String.format("SELECT * FROM %s WHERE METRIC_NAME LIKE '%s' AND SERVER_TIME > %s ORDER BY METRIC_NAME, SERVER_TIME", inputTable, metricName, startTime);
+        runPhoenixQueryAndAddToResults(query);
```

Comment:
The implementation flow looks like this now (Please correct me if I am wrong).
- Read all metrics from old tables
- Write all metrics to new tables
- Save progress

I was hoping it would be like this.
For every iteration,
- Pick a small batch of metrics.
- Read from old table
- Write to new table
- Save progress

Reply:
This will be configured with batch size, once implemented, we will pass the batch to each thread to work with